<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>JavaScript - Asynchronous JavaScript and APIs - Async/Await Tasks</title>
</head>
<body>
    <script>
        // Task 1 - Rewrite using async/await
        (() => {
            async function loadJson(url) {
                let response = await fetch(url);
                if (response.status == 200) {
                    return await response.json();
                }
                // If reach here, throw error
                throw new Error(response.status);
            }

            loadJson('https://javascript.info/no-such-user.json')
                .catch(alert); // Error: 404
        })();

        // Task 2 - Rewrite "rethrow" with async/await
        (() => {
            class HttpError extends Error {
                constructor(response) {
                    super(`${response.status} for ${response.url}`);
                    this.name = 'HttpError';
                    this.response = response;
                }
            }

            async function loadJson(url) {
                let response = await fetch(url);

                if (response.status == 200) {
                    return await response.json();
                }

                throw new HttpError(response);
            }

            async function demoGithubUser() {
                while (true) {
                    let name = prompt("Enter a name?", "iliakan");
                    
                    try {
                        let user = await loadJson(`https://api.github.com/users/${name}`);
                        alert(`Full name: ${user.name}`);
                        return user;
                    } catch (err) {
                        if (err instanceof HttpError && err.response.status == 404) {
                            alert("No such user, please reenter.");
                        } else {
                            throw err;
                        }
                    }
                }
            }
            
            demoGithubUser();
        })();

        // Task 3 - Call async from non-async
        (() => {
            async function wait() {
                await new Promise(resolve => setTimeout(resolve, 1000));

                return 10;
            }

            function f() {
                wait()
                    .then(response => {
                        alert(response);
                    })
                    .catch(err => {
                        alert(err);
                    });
            }

            f();

        })();
    </script>
</body>
</html>